src/core/namespace.c

```diff
@@ -519,6 +539,33 @@ static int append_extensions(
                                         "No matching entry in .v/ directory %s found.",
                                         m->source);
 
+                /* If this is a dm-verity image, check for duplicates, as Overlayfs rejects them as
+                 * explained above. */
```

Comment:
this is really misleading, as mentioned.

it's hard to dedup this correctly from userspace, but what you have to do goes something like this:

1. for each layer read the uuid off the superblock via FS_IOC_GETFSUUID. If that doesn't work, use fstatfs() instead and use the `.fs_id` field instead.
2. use name_to_handle_at() on the top-level inode, and use the new AT_HANDLE_FID flag. 

If both the uuid/fs_id and the filehandle blob of the top-level inode match then this filter one of the two layers.

if you reference the same inode on two dm-verity layers which have the same roothash then you *of* *course* will also pass the test above, but that's a much more specific check.

Reply:
Not sure it is _that_ misleading, given the issue is using the same device mapper node - but this approach is enough to make it work, what is the advantage of doing all that extra work? If the roothash is identical, it's cryptographically guaranteed to be the same. If it's different, it's guaranteed to be different. I'm not aware of any possibility of having false positives/negatives?

Reply:
As far as I can tell there's also a major problem with that suggestion: it requires the filesystem to be _already_ mounted, but in namespace.c we do all the checks and the deduping in the first step, where everything is prepared, and the mounting after, when everything is ready. Checking the actual superblock metadata requires mounting first, no? Which would break this pattern - suddenly for this one type, the actual things mounted can change after all the mounts are decided and prepared, instead of before.

Reply:
i do accept that filtering before verity does make sense.